src/main/java/hudson/remoting/JarCacheSupport.java

```diff
@@ -50,50 +51,92 @@
-                        } catch (ChannelClosedException e) {
-                            // the connection was killed while we were still resolving the file
-                            bailout(e);
-                        } catch (RequestAbortedException e) {
-                            // the connection was killed while we were still resolving the file
-                            bailout(e);
-                        } catch (InterruptedException e) {
-                            // we are bailing out, but we need to allow another thread to retry later.
-                            bailout(e);
+                try {
+                    final AsyncFutureImpl<URL> promise = new AsyncFutureImpl<URL>();
+                    ExecutorServiceUtils.submitAsync(downloader, new  DownloadRunnable(channel, sum1, sum2, key, promise));
+                    // Now we are sure that the task has been accepted to the queue, hence we cache the promise
+                    // if nobody else caches it before.
+                    inprogress.putIfAbsent(key, promise);
+                } catch (ExecutorServiceUtils.ExecutionRejectedException ex) {
+                    final String message = "Downloader executor service has rejected the download command for checksum " + key;
+                    LOGGER.log(Level.SEVERE, message, ex);
+                    // Retry the submission after 100 ms if the error is not fatal
+                    if (ex.isFatal()) {
+                        // downloader won't accept anything else, do not even try
+                        throw new IOException(message, ex);
+                    } else {
+                        //TODO: should we just fail? unrealistic case for the current AtmostOneThreadExecutor implementation anyway
+                        Thread.sleep(100);
+                    }
+                }
+            }
+        }
+    }
+    
+    private class DownloadRunnable implements Runnable {
+    
+        final Channel channel;
+        final long sum1;
+        final long sum2;
+        final Checksum key;
+        final AsyncFutureImpl<URL> promise;
 
-                            LOGGER.log(Level.WARNING, String.format("Interrupted while resolving a jar %016x%016x",sum1,sum2), e);
-                        } catch (Throwable e) {
-                            // in other general failures, we aren't retrying
-                            // TODO: or should we?
-                            promise.set(e);
+        public DownloadRunnable(Channel channel, long sum1, long sum2, Checksum key, AsyncFutureImpl<URL> promise) {
+            this.channel = channel;
+            this.sum1 = sum1;
+            this.sum2 = sum2;
+            this.key = key;
+            this.promise = promise;
+        }
+        
+        @Override
+        public void run() {
+            try {
+                // Deduplication: There is a risk that multiple downloadables get scheduled, hence we check if
```

Comment:
The logic has been moved to a nested class, The deduplication section below is the only difference